kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/message/store/SqlMessageStoreHelper.java

```diff
@@ -0,0 +1,330 @@
+            throws KuraStoreException {
+
+        return this.connectionProvider.withPreparedStatement(this.queries.getSqlGetMessage(), (c, stmt) -> {
+            stmt.setInt(1, msgId);
+
+            return getFirstColumnValueOrEmpty(stmt::executeQuery, (rs, i) -> messageBuilder.call(rs));
+
+        }, "Cannot get message by ID: " + msgId);
+    }
+
+    public Optional<StoredMessage> getNextMessage() throws KuraStoreException {
+
+        return getNextMessage(rs -> buildStoredMessageBuilder(rs, true).build());
+
+    }
+
+    public Optional<StoredMessage> getNextMessage(final SQLFunction<ResultSet, StoredMessage> messageBuilder)
+            throws KuraStoreException {
+
+        return this.connectionProvider.withPreparedStatement(this.queries.getSqlGetNextMessage(),
+                (c, stmt) -> getFirstColumnValueOrEmpty(stmt::executeQuery, (rs, i) -> messageBuilder.call(rs)),
+                "Cannot get message next message");
+    }
+
+    public void markAsPublished(int msgId, DataTransportToken token) throws KuraStoreException {
+        final Timestamp now = new Timestamp(new Date().getTime());
+
+        this.connectionProvider.withPreparedStatement(this.queries.getSqlSetPublishedQoS1(), (c, stmt) -> {
+
+            stmt.setTimestamp(1, now, this.utcCalendar);
+            stmt.setInt(2, token.getMessageId());
+            stmt.setString(3, token.getSessionId());
+            stmt.setInt(4, msgId);
+
+            stmt.execute();
+
+            if (isExplicitCommitEnabled) {
+                c.commit();
+            }
+            return null;
+
+        }, "Cannot update timestamp");
+
+    }
+
+    public void markAsPublished(int msgId) throws KuraStoreException {
+        updateTimestamp(this.queries.getSqlSetPublishedQoS0(), msgId);
+    }
+
+    public void markAsConfirmed(int msgId) throws KuraStoreException {
+        updateTimestamp(this.queries.getSqlSetConfirmed(), msgId);
+    }
+
+    public List<StoredMessage> getUnpublishedMessages() throws KuraStoreException {
+
+        return listMessages(this.queries.getSqlAllUnpublishedMessages());
+    }
+
+    public List<StoredMessage> getInFlightMessages() throws KuraStoreException {
+
+        return listMessages(this.queries.getSqlAllInFlightMessages());
+    }
+
+    public synchronized List<StoredMessage> getDroppedMessages() throws KuraStoreException {
+
+        return listMessages(this.queries.getSqlAllDroppedInFlightMessages());
+    }
+
+    public void unpublishAllInFlighMessages() throws KuraStoreException {
+        execute(this.queries.getSqlUnpublishAllInFlightMessages());
+    }
+
+    public void dropAllInFlightMessages() throws KuraStoreException {
+        updateTimestamp(this.queries.getSqlDropAllInFlightMessages());
+    }
+
+    public void deleteStaleMessages(int purgeAge) throws KuraStoreException {
+        final long timestamp = System.currentTimeMillis() - TimeUnit.SECONDS.toMillis(purgeAge);
```

Comment:
A subtle thing can happen here: the long will be set as long in the underlying `setObject()` of the execute method. Some databases might not automatically reparse this as a `Timestamp`. So I would replace this line with:

```java
final long difference = System.currentTimeMillis() - TimeUnit.SECONDS.toMillis(purgeAge);
Timestamp timestamp = new Timestamp(difference);
```